### PR TITLE
Kinetis Replace DMA with a proper eDMA driver and use it on SPI and Serial

### PR DESCRIPTION
**Pull request by davids5**
## Summary

This PR replaces the DMA driver from 2017 that was not very useful, with a port in of the S32K eDMA driver. Several fixes were needed to get it to build and run correctly. One simple enhancement was added for destination looping with only one channel/csd to facilitate a simple circular buffer. 

The serial driver was updated to use the new interface.
The Serial driver no longer needs a DMA poll as it is driven off of RX idle.
Added SPI DMA using the new interface.
    Kinetis:SPI only allocate DMA once

## Impact

None in tree.

## Testing

PX4 nxp_fmuk66-e

### CODE REVIEW

File: arch/arm/src/kinetis/Kconfig [hunk 1]
```diff
@@ -1114,6 +1114,96 @@ config KINETIS_SD4BIT_FREQ
...
+		Normally TCD word 2 is a 32-bit NBYTES field.  When this option is
+		enabled, TCD word 2 is redefined to include individual enable fields,
+		an offset field, and the NBYTES field.  The individual enable fields
+		allow the minor loop offset to be applied to the source address, the
+		destination address, or both. The NBYTES field is reduced when either
+		offset is enabled.
+
+config KINETIS_EDMA_EDBG
+	bool "Enable Debug"
+	default n
```
Comment (acassis) [inline arch/arm/src/kinetis/Kconfig:1204]: Shouldn't it depend on DEBUG being enabled?

File: arch/arm/src/kinetis/kinetis_edma.h [hunk 2]
```diff
@@ -0,0 +1,455 @@
+/****************************************************************************
+ * arch/arm/src/kinetis/kinetis_edma.h
+ *
+ *   Copyright (C) 2019 Gregory Nutt. All rights reserved.
+ *   Author: Gregory Nutt <gnutt@nuttx.org>
+ *
+ * This file was leveraged from the NuttX i.MXRT port.
+ * Portions of that eDMA logic derived from NXP sample code which has
+ * a compatible BSD 3-clause license:
+ *
```
Comment (acassis) [inline arch/arm/src/kinetis/kinetis_edma.h:10]: Please include the Apache License here and keep the Freescale/NXP BSD License below. This is how Samsung did when they released TIzenRT as Apache using NuttX BSD License.

File: arch/arm/src/kinetis/kinetis_edma.c [hunk 3]
```diff
@@ -0,0 +1,1464 @@
+/****************************************************************************
+ * arch/arm/src/kinetis/kinetis_edma.c
+ *
+ *   Copyright (C) 2019, 2021 Gregory Nutt. All rights reserved.
+ *   Authors: Gregory Nutt <gnutt@nuttx.org>
+ *            David Sidrane <david.sidrane@nscdg.com>
+ *
```
Comment (acassis) [inline arch/arm/src/kinetis/kinetis_edma.c:7]: Please include the Apache License here and keep the Freescale/NXP BSD License below. This is how Samsung did when they released TIzenRT as Apache using NuttX BSD License.

File: arch/arm/src/kinetis/Kconfig (hunk 1 above)
Comment (davids5) [reply arch/arm/src/kinetis/Kconfig:1204]: No. This is the thing the allows the DMA to be stopped while stopped in a JTag.

File: arch/arm/src/kinetis/kinetis_edma.h (hunk 2 above)
Comment (davids5) [reply arch/arm/src/kinetis/kinetis_edma.h:10]: @acassis do we have a CCLA from NXP? I thought this was still tied up and should not be changed until it was resolved.

File: arch/arm/src/kinetis/kinetis_edma.c (hunk 3 above)
Comment (davids5) [reply arch/arm/src/kinetis/kinetis_edma.c:7]: "

File: arch/arm/src/kinetis/kinetis_spi.c [hunk 4]
```diff
@@ -1251,6 +1664,39 @@ FAR struct spi_dev_s *kinetis_spibus_initialize(int port)
...
   nxsem_init(&priv->exclsem, 0, 1);
+#ifdef CONFIG_KINETIS_SPI_DMA
+  /* Initialize the SPI semaphores that is used to wait for DMA completion.
+   * This semaphore is used for signaling and, hence, should not have
+   * priority inheritance enabled.
+   */
+
+  if (priv->rxch && priv->txch)
+    {
+      if (priv->txdma == NULL && priv->rxdma == NULL)
```
Comment (davids5) [inline arch/arm/src/kinetis/kinetis_spi.c:1675]: @acassis - this was the last force push. Kinetis:SPI only allocate DMA once

**Review by xiaoxiang781216 — APPROVED** [review]

**Pull request merged**